main/src/ca/uwaterloo/flix/api/Bootstrap.scala

```diff
@@ -424,6 +426,102 @@ class Bootstrap(val projectPath: Path, apiKey: Option[String]) {
     }
   }
 
+  /**
+    * Sequentially deletes all `.class` files in the build directory if ALL files (recursively) in the build directory
+    * are `.class` files.
+    * Aborts if any other file was found.
+    */
+  def clean()(implicit formatter: Formatter): Result[Unit, BootstrapError] = {
+    // Ensure project mode
+    if (optManifest.isEmpty) {
+      return Err(BootstrapError.FileError("No manifest found. Run 'flix init' to set up project mode."))
+    }
+
+    val buildDir = Bootstrap.getBuildDirectory(projectPath)
+
+    checkForDangerousPath(buildDir) match {
+      case Err(e) => return Err(e)
+      case Ok(()) => ()
+    }
+
+    // Ensure all files in `buildDir` are valid class files.
+    val paths = FileOps.getFilesIn(buildDir, Int.MaxValue)
+    for (path <- paths) {
+      if (!FileOps.checkExt(path, "class")) {
+        return Err(BootstrapError.FileError(s"Unexpected file extension in build directory (only '.class' files are allowed): '${buildDir.relativize(path)}'"))
+      }
+
+      if (!FileOps.isClassFile(path)) {
+        return Err(BootstrapError.FileError(s"Invalid class file in build directory: '${buildDir.relativize(path)}'"))
+      }
+    }
+
+    // Collect dirs in list, so they can be deleted after deleting all their contents.
+    val dirs = mutable.ArrayBuffer.empty[Path]
+
+    // Delete non-directory paths
+    for (path <- paths) {
+      if (Files.isDirectory(path)) {
+        dirs.addOne(path)
+      } else {
+        checkForDangerousPath(path) match {
+          case Err(e) => return Err(e)
+          case Ok(()) => ()
+        }
+
+        FileOps.delete(path) match {
+          case Err(e) => return Err(BootstrapError.FileError(e.getMessage))
+          case Ok(_) => ()
+        }
+      }
+    }
+
+    // Delete empty directories
+    for (dir <- dirs) {
+      checkForDangerousPath(dir) match {
+        case Err(e) => return Err(e)
+        case Ok(()) => ()
+      }
+
+      FileOps.delete(dir) match {
+        case Err(e) => return Err(BootstrapError.FileError(e.getMessage))
+        case Ok(_) => ()
+      }
+    }
+
+    Ok(())
+  }
+
+  /**
+    * Returns `Err` if `path` is one of the follwing:
+    *   1. The root directory of the system
+    *   1. The user's home directory (`~/`)
+    *   1. Any ancestor of [[projectPath]]
+    *
+    * Returns `Ok(())` otherwise.
+    */
+  private def checkForDangerousPath(path: Path)(implicit formatter: Formatter): Result[Unit, BootstrapError] = {
+    val root = Path.of("/").normalize()
+    val home = Path.of("~/").normalize()
+
+    // Ensure `buildDir` is NOT the root directory.
+    if (root == path) {
+      return Err(BootstrapError.FileError(formatter.red("Found root directory. Aborting...")))
```

Comment:
We did not use the formatter elsewhere. Perhaps we leave that out for now. Or at least be consistent. Usually we should also have specific error messages.